File: 6.Sorted Challenges  310p/4.二叉树与递归 38/遍历器相关/173. Binary Search Tree Iterator.ts
namespace LeetCode173 {
	class BSTIterator {
		idx: number
		arr: number[]
		constructor(root: TreeNode) {
			this.idx = 0
			this.arr = []
			this.getNode(root)
		}
		getNode = (node: TreeNode) => {
			if (!node) return
			this.getNode(node.left!)
			this.arr.push(node.val)
			this.getNode(node.right!)
		}
		next = () => {
			return this.arr[this.idx++]
		}
		hasNext = () => {
			if (this.idx >= this.arr.length) return false
			else return true
		}
	}

	class TreeNode {
		val: number
		left: TreeNode | null
		right: TreeNode | null
		constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
			this.val = val === undefined ? 0 : val
			this.left = left === undefined ? null : left
			this.right = right === undefined ? null : right
		}
	}
}

File: legacy/6.Sorted Challenges  310p/4.二叉树与递归 38/遍历器相关/173. Binary Search Tree Iterator.java
import java.util.LinkedList;

class BSTIterator {
  private int idx;
  private List<Integer> arr;

  public BSTIterator(TreeNode root) {
    idx = 0;
    arr = new ArrayList<Integer>();
    traverse(root, arr);
  }

  public int next() {
    return arr.get(idx++);
  }

  public boolean hasNext() {
    return idx < arr.size();
  }

  private void traverse(TreeNode root, List<Integer> arr) {
    if (root == null) {
      return;
    }
    traverse(root.left, arr);
    arr.add(root.val);
    traverse(root.right, arr);
  }
}